### DISCUSSION TITLE
Batch Receive

### DISCUSSION THREAD

**Discussion by oguzhantortop** [majusko/pulsar-java-spring-boot-starter#286] [Ideas]
Can we receive messages in batches? Normally pulsar has the ability to receive batch messages in a given quantity in a given timeout. If a topic does not receive enough messages then it consumes messages in a given time. Does a reactive client of yours works in this fashion or do you need additional development for this feature?

**Comment by majusko** [comment]
Yes, that is how the Pulsar works, but the configuration is kept on default values. Would you like to have it exposed to override batching policy?

**Reply by majusko** [reply]
```
    public static final BatchReceivePolicy DEFAULT_POLICY = new BatchReceivePolicy(
            -1, 10 * 1024 * 1024, 100, TimeUnit.MILLISECONDS)
```

**Reply by oguzhantortop** [reply]
Not quite sure if we are talking about the same thing. I was actually trying to say that pulsar consumer has a method called consumer.batchReceive(). This is effective for scenarios like batch inserting to database. I looked at your code a little bit the one with PulsarConsumer annotation and as far as I see it doesn't use consumer.batchReceive() instead it uses messageListener. And in that case you might receive msgs in batch but you don't process them in batch... I hope I can express it...

**Reply by majusko** [reply]
I see. Thanks for explaining the DB use case. You are right, this is a good idea to implement in the library. Thank you 🙌

**Reply by oguzhantortop** [reply]
If I will have some time I would like to contribute. Though I am not familiar with Spring Reactive elements I might help for the non reactive part. A new annotation like @PulsarBatchConsumer might be good...

**Reply by majusko** [reply]
Yes, definitely should be a separate annotation. However, this would be quite a lot of work as you would need to have a different subscription mechanism since after the batchReceive timeout or msg limit is reached, the subscription needs to be renewed. Also, many tests would need to be configured there, so please go ahead only if you are confident about how to implement it. I may also help with the implementation. 🤞

**Comment by oguzhantortop** [comment]
What I am thinking is to create an infinite loop that calls the consumer batchReceive method. And this infinite loop should run inside a separate thread or maybe we might use another feature of the spring framework. I will start to write but I am not sure if this is the right approach. What do you think about it?

**Comment by oguzhantortop** [comment]
Instead creating a new annotation I used your existing annotation by adding a batch attribute. default value is false when you set it to true than I iniate a new thread with infinite loop. Of course there are tons of things to do. I just committed it for you to take a look at it. 
https://github.com/oguzhantortop/pulsar-java-spring-boot-starter btw I switched to pulsar 2.9.2 because of my current environment...

**Reply by majusko** [reply]
Yes, I was also thinking of this, and you are right. This should work perfectly as well and shouldn't be that complicated to implement in the end. I think you can continue with this, adding exception handling, cleaning the code a bit and adding a few test cases for sending messages in batches. Make sure to log errors and send them also into the `sink`. Would be good to update a README and probably even describe the case in the commentary of `@interface PulsarConsumer` to explain the type difference.

Nice work! 🙌

**Reply by oguzhantortop** [reply]
What do you think about ack? I think there could be an attribute about manual ack control and the default value could be auto. In the auto ack mode, the system would ack whole msgs if everything goes well without any exception. if an exception happens while processing one of the msg whole msgs should be negatively acknowledged. Or we may request the user to return a list of msg ids to be acked but in this case, the user should do exception control on its own.   But within manual control then we should also inject consumer object into the user's method. 

A real theoric problem... We have to keep it simple...

**Reply by majusko** [reply]
Within this PR, we should do a more straightforward solution and therefore implement automatic acknowledgement and negatively acknowledge all messages on an exception that occurred, as you described. 

I would separate the work for manual handling into separate PR and add this option to standard single msg consuming as well later on.

Your PR should also add a batching configuration(policy), I believe, and it should also be inside the annotation where -1 would mean default configuration. I think it's only 2 configs - timeout and number of msgs.

**Reply by oguzhantortop** [reply]
I have added a batchAckControl attribute default value is "auto" and it only works when batch attribute is set to true. And tried to summarize possible usage. But still I feel the code became little bit complicated, I also committed the changes. Could you please look at code and possibilities? 

 Possible Scenarios:
1) user sets batch true and sets batchAckControl attribute to auto or doesn't uses at all.
    - if no error arises during msgs consume or if user swallow exception then we auto-ack all the messages.
    - if any exception occurs then we negative ack all the messages. (Also we could implement deadqueue and redelivery count logic in here as well.
 ```
public void consumeString(Message<byte[]> msgs) {
    }
```
2)  user sets batch true and sets batchAckControl attribute to auto or doesn't uses at all. But this time users method has a return type of List<MessageId>. 
      - if no error arises during msgs consume or if user swallow exception then we acknowledge only returned List of messages. We might need another user to return in order to negative-ack rest of the msgs, or we might filter rest of the msgs to give negative ack.
      - If any exception occurs we negative ack all the messages.
 ```
public List<MessageId> consumeString(Message<byte[]> msgs) {
    }
```
3) user sets batch true and sets batchAckControl attribute to manual
     - This time we expect user to create a method with two parameters. The first one is type of Messages<?> class of Pulsar and the second one is Consumer object. We need consumer object in order to give ack or negative ack... Also we could negative ack if an exception occurs from the users method...
```
     public void consumeString(Messages<byte[]> msgs,Consumer<byte[]> consumer) {
    		try {
    			List<MessageId> list = new ArrayList<>();
    			List<MappedObj> dbList = new ArrayList<>();
    			msgs.forEach((msg) -> {
    				try {
    					MappedObj tmp = Mapper.jsonToBean(new String(msg.getValue()));
        				dbList.add(tmp);
        				list.add(msg.getMessageId());
					} catch (Throwable e) {
						System.err.println("Error while processing msg: "+String.valueOf(msg.getValue())+ " error msg: "+e.getMessage());
					}
    			});
    			eventSearchRepo.saveAll(dbList);
    			consumer.acknowledge(list);
			} catch (Throwable e) {
				System.err.println(e.getMessage());
			}
    }
```

**Reply by majusko** [reply]
Yes, this seems like a good design. I think you should probably change one path:

```
} else if(!retTypeVoid) {
   ackList = (List<MessageId>) method.invoke(holder.getBean(), msgs);
   consumer.acknowledge(ackList);
   
   // you should filter msgs and negatively ack those missing in the list
   
} else {
```

**Reply by oguzhantortop** [reply]
I completed filtering could you pls have a look when you have time.

**Reply by majusko** [reply]
Yes, logic is good - one last null pointer I see there:
```
} catch (Exception e) {
    if(retTypeVoid && !manualAckMode) {
	    consumer.negativeAcknowledge(msgs); // msgs can be null - please do a check for null as well inside an IF 
    } 
}
```

 Regarding formatting, I would like you to apply - also, make sure to check for future changes:

```
List<MessageId> ackList = null; // this can be final iitialised inside else
```


```
ackSet // final
```

- Please check the formatting - you are using 8 whitespaces for some type of separation

- Please use curled braces for `if(` statements. I would be ok with this in Kotlin but not in Java.

- Don't use magic strings - `"manual"` - put it into some constant, please.

**Reply by oguzhantortop** [reply]
Hi, I fixed above findings of yours. Tried to reformat code with eclipse not quite sure if I did as you expected. If there are still unwanted chars could you pls give line col numbers of them? I will keep add comments, logging and sink as you wanted. Also I expect your comments about CompletableFuture usage and threadpool usage. 

thanks

**Reply by oguzhantortop** [reply]
Hi,
I have added documentation for annotation attributes and fixed the magic strings problems and some code smell. Could you please have a look?

Also, we need to set batch receive policies. Should we put those values from annotation attributes or should we use a properties file?

**Reply by majusko** [reply]
Thank you, @oguzhantortop Much better now. 

Few points:

1. You are swallowing exceptions there. You forgot to send all exceptions to the `sink` so users can subscribe on all errors happening during the subscription:

```
sink.tryEmitNext(new FailedMessage(e, consumer, msg));
```

2.

```
msgs != null // can be connected with previous if statement

```

3. Please collect completable futures as we do with `private List<Consumer> consumers;` just adding into a list + creating a getter inside a class so people can debug the thread's state if needed.


4. try to separate the whole batching part of the method to separate method because it's quite a robust method now
 from `if (holder.getAnnotation().batch()) {` until the end of the if
 
 5. please configure your formatting on 4-spaces tabs - you have it at 8 - I'm not using Eclipse, but it should be pretty easy to configure for the project - In IntelliJ, it would be under the Code Style part in preferences. It also seems you are using tabs instead of whitespace. Also, make sure you don't change part of code that was developed by other programmers.
 
 6. Add Policy - just use annotation fields - no need to put global configuration since there is always a default policy.
 
 7. Add a few test cases
 
 8. Update a README.md
 
Once all of these points are solved, please feel free to create PR - it's much easier to review that way, and I believe there might be only a few last cosmetics, and then this can be merged. Really, excellent work!

**Reply by oguzhantortop** [reply]
For the 1. item FailedMessage class is not suitable with Messages class. So I created a new class like below:
    `private final Sinks.Many<FailedMessages> batchSink = Sinks.many().multicast().onBackpressureBuffer(Queues.SMALL_BUFFER_SIZE, false);`

I will do the final formatting on IntelliJ because my Eclipse shows 4 spaces tab already anyway I tried to do it manually as much as possible. I completed items other than 7. and 8. After completing 7 and 8 I will create PR. Till that time if something comes to your mind please let me know...

**Reply by majusko** [reply]
I think it is better to have one sink for errors, so it's easier to write loggers and people that rely on this already don't need to change anything. So please extend the class used in the sink with the new attribute `batchMessages` with a new getter and keep everything as it is so it's not a breaking change.

**Reply by majusko** [reply]
Sorry for the formatting - maybe it's just a GitHub UI - will check that in IDE once you raise a PR.

**Comment by oguzhantortop** [comment]
Added documentation to batch consumer configuration. Planning to complete unit tests this weekend.

**Comment by oguzhantortop** [comment]
Hi, I also finished with unit tests.

**Comment by majusko** [comment]
Perfect, one last thing - 2 tests are failing - the hardcoded consumers and producers calculator. Please increase the values. Also wrote a comment to the PR. After this I will merge it :)

**Comment by oguzhantortop** [comment]
Can you control again? :)

**Reply by majusko** [reply]
Done, merged! 💯 Congrats 🎉 ... I will do a release soon as some other minor changes should go out asap.

**Comment by riritack** [comment]
How to control the time interval of each batch in batch consume

**Reply by oguzhantortop** [reply]
you need to set timeoutMillis attribute of consumer annotation. But as far as I know it is available on version 1.2.0.

**Reply by riritack** [reply]
ok,thanks
